### PR TITLE
psst: init at 318e069

### LINKED ISSUE #115155: package request: psst - "Native Spotify client for the Gnome desktop"
**Project description**

>Fast and multi-platform Spotify client with native GUI

>Fast Spotify client with native GUI, without Electron, built in Rust. Very early in development, lacking in features, stability, and general user experience. It is being tested only on Mac so far, but aims for full Windows and Linux support. Contributions welcome!

**Metadata**

* homepage URL: https://github.com/jpochyla/psst
* source URL: https://github.com/jpochyla/psst
* license: **mit**
* platforms: **darwin, linux**

### PR DESCRIPTION
**Pull request by matthias-t**
###### Motivation for this change

Fixes #115155. This won't build until `rustc` 1.54.0 is merged.

Tangentially, I resorted to `fetchgit` since `fetchFromGitHub` seemingly fails to fetch the minivorbis submodule, even with `fetchSubmodules = true`. Could somebody reproduce or investigate this?

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/psst/default.nix [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  cargoSha256 = "1f5cn0ncwj869snil4fid2ybm09dngj13ls5h0wbqp9ypk9097bq";
+
+  meta = {
+    description = "Fast and multi-platform Spotify client with native GUI";
+    homepage = "https://github.com/jpochyla/${pname}";
+    license = lib.licenses.mit;
+    platforms = lib.platforms.all;
+  };
+}
+
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/psst/default.nix:30]: Final new line to much.

File: pkgs/applications/audio/psst/default.nix [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  nativeBuildInputs = [ pkg-config ];
+  buildInputs = [ dbus ];
+
+  cargoSha256 = "1f5cn0ncwj869snil4fid2ybm09dngj13ls5h0wbqp9ypk9097bq";
+
+  meta = {
+    description = "Fast and multi-platform Spotify client with native GUI";
+    homepage = "https://github.com/jpochyla/${pname}";
+    license = lib.licenses.mit;
+    platforms = lib.platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/psst/default.nix:27]:
```suggestion
  meta = with lib; {
    description = "Fast and multi-platform Spotify client with native GUI";
    homepage = "https://github.com/jpochyla/psst";
    license = licenses.mit;
    platforms = platforms.all;
```
Missing maintainer.

File: pkgs/applications/audio/psst/default.nix [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+let
+  rev = "318e069ca85d0f58358d9bc8d879ecdd4d4ec0ea";
+in
+
+rustPlatform.buildRustPackage rec {
+  pname = "psst";
+  version = builtins.substring 0 7 rev;
+
+  src = fetchgit {
+    url = "https://github.com/jpochyla/${pname}";
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/psst/default.nix:12]:
```suggestion
    url = "https://github.com/jpochyla/psst";
```
Stop substituting every occurrences of pname and use fetchFromGitHub.

File: pkgs/applications/audio/psst/default.nix [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{ lib, rustPlatform, fetchFromGitHub, pkg-config, dbus, fetchgit, ... }:
+
+let
+  rev = "318e069ca85d0f58358d9bc8d879ecdd4d4ec0ea";
+in
+
+rustPlatform.buildRustPackage rec {
+  pname = "psst";
+  version = builtins.substring 0 7 rev;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/psst/default.nix:9]: Use unstable-XXXX-XX-XX format.

File: pkgs/applications/audio/psst/default.nix [hunk 5]
```diff
@@ -0,0 +1,30 @@
+{ lib, rustPlatform, fetchFromGitHub, pkg-config, dbus, fetchgit, ... }:
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/psst/default.nix:1]:
```suggestion
{ lib, rustPlatform, fetchFromGitHub, pkg-config, dbus, fetchgit }:
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by figsoda** [comment]
> Tangentially, I resorted to `fetchgit` since `fetchFromGitHub` seemingly fails to fetch the minivorbis submodule, even with `fetchSubmodules = true`. Could somebody reproduce or investigate this?

I can't reproduce this. I tested with this:

```nix
(import <nixpkgs> {}).fetchFromGitHub {
  owner = "jpochyla";
  repo = "psst";
  rev = "318e069ca85d0f58358d9bc8d879ecdd4d4ec0ea";
  sha256 = "0i18i4d1l0vkzz21jwfxhshp3p0w5q8q1dlxlf39b0jjxsgx78qk";
  fetchSubmodules = true;
}
```

File: pkgs/applications/audio/psst/default.nix [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  nativeBuildInputs = [ pkg-config ];
+  buildInputs = [ dbus ];
+
+  cargoSha256 = "0z8hwd223sgsf1kp6q84rbhh3ri28jbqr2zzg5hn8arbdif3i2bn";
+
+  meta = with lib; {
+    description = "Fast and multi-platform Spotify client with native GUI";
+    homepage = "https://github.com/jpochyla/psst";
+    license = licenses.mit;
+    platforms = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/psst/default.nix:24]:
```suggestion
```

ste by buildRustPackage

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

**Comment by SuperSandro2000** [comment]
> Tangentially, I resorted to fetchgit since fetchFromGitHub seemingly fails to fetch the minivorbis submodule, even with fetchSubmodules = true. Could somebody reproduce or investigate this?

You need to invalidate the hash.

**Comment by matthias-t** [comment]
Yes, OfBorg runs into [the same](https://logs.nix.ci/?key=nixos/nixpkgs.134921&attempt_id=33707804-a0a7-4de8-bbdc-eec9590ba20a). That's odd, for me the fetcher seems to have succeeded:

<details>
<summary><code>nix-build -A psst</code></summary>
<pre lang="console">
these derivations will be built:
  /nix/store/0v062f8dszz1m7hhm6r0ln5wys1x2g09-psst-unstable-2021-08-19.drv
building '/nix/store/0v062f8dszz1m7hhm6r0ln5wys1x2g09-psst-unstable-2021-08-19.drv'...
unpacking sources
unpacking source archive /nix/store/487dnkc4zgxzqy26pz022yv9ac0dq294-source
source root is source
Executing cargoSetupPostUnpackHook
unpacking source archive /nix/store/mp39hnipp70i4l9avbdzasby2hlkby0j-psst-unstable-2021-08-19-vendor.tar.gz
Finished cargoSetupPostUnpackHook
patching sources
Executing cargoSetupPostPatchHook
Validating consistency between /build/source//Cargo.lock and /build/psst-unstable-2021-08-19-vendor.tar.gz/Cargo.lock
Finished cargoSetupPostPatchHook
configuring
building
Executing cargoBuildHook
++ env CC_x86_64-unknown-linux-gnu=/nix/store/4xs1xyj8728yvh9y5v6ji819kwgfy2fv-gcc-wrapper-10.3.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/4xs1xyj8728yvh9y5v6ji819kwgfy2fv-gcc-wrapper-10.3.0/bin/c++ CC_x86_64-unknown-linux-gnu=/nix/store/4xs1xyj8728yvh9y5v6ji819kwgfy2fv-gcc-wrapper-10.3.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/4xs1xyj8728yvh9y5v6ji819kwgfy2fv-gcc-wrapper-10.3.0/bin/c++ cargo build -j 4 --target x86_64-unknown-linux-gnu --frozen --release
   Compiling proc-macro2 v1.0.28
   Compiling unicode-xid v0.2.2
   Compiling syn v1.0.74
   Compiling serde v1.0.127
   Compiling serde_derive v1.0.127
[...]
</pre>
</details>

```console
$ nix-store -qR /nix/store/0v062f8dszz1m7hhm6r0ln5wys1x2g09-psst-unstable-2021-08-19.drv
[...]
/nix/store/1270kx4q69baf21zagcdsjksc402rfjz-psst-unstable-2021-08-19-vendor.tar.gz.drv
[...]
```

```console
$ nix-store -q /nix/store/1270kx4q69baf21zagcdsjksc402rfjz-psst-unstable-2021-08-19-vendor.tar.gz.drv
/nix/store/mp39hnipp70i4l9avbdzasby2hlkby0j-psst-unstable-2021-08-19-vendor.tar.gz
```

**Comment by fabianhjr** [comment]
Do you have sandboxing enabled?

**Comment by matthias-t** [comment]
I do, according to `nix-info -m`.

**Comment by ethancedwards8** [comment]
Any progress on this?

**Comment by matthias-t** [comment]
The fetch didn't fail because of sandboxing, it failed because the remote couldn't find the commit. I'm not sure why exactly nix was attempting to fetch commit `30aa0baa` on the `psst` branch of the druid repo, but it [appears](https://github.com/jpochyla/druid/commit/30aa0baa) to no longer be part of the branch, supposedly after a force-push. Whatever may have occured exactly, the build should now succeed. However, the application fails for me on wayland with a blank small square window and:

```
(psst-gui:141599): Gtk-CRITICAL **: 12:05:18.310: gtk_window_set_default_size: assertion 'width >= -1' failed
```

**Comment by glittershark** [comment]
Result of `nixpkgs-review pr 134921` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>psst</li>
  </ul>
</details>

**Comment by glittershark** [comment]
^ that was with sandboxing enabled, fwiw

**Comment by figsoda** [comment]
looks like darwin build needs: AppKit CoreFoundation CoreGraphics CoreText CoreVideo Foundation MediaPlayer QuartzCore Security

aarch64-linux build also failed, I don't know what the error messages mean, maybe it just doesn't support it

File: pkgs/applications/audio/psst/default.nix [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    rev = "724aa251407dfa93b8ede8e879fd5ca35de1ddd2";
+    sha256 = "08m6h2gkgnpmfpw93zb6rhrvbw5rhq0mgpcfzingdvsqw531r13w";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [ pkg-config ];
+  buildInputs = [ gtk3 glib gdk-pixbuf cairo pango dbus ]
+  ++ lib.optionals stdenv.isDarwin (with darwin.apple_sdk.frameworks; [
+    AppKit CoreFoundation CoreGraphics CoreText CoreVideo
+    Foundation MediaPlayer QuartzCore Security ]);
```
Comment (figsoda) [inline pkgs/applications/audio/psst/default.nix:20]: can you please inherit these inside `all-packages.nix`?
```suggestion
    ++ lib.optionals stdenv.isDarwin [
      AppKit CoreFoundation CoreGraphics CoreText CoreVideo
      Foundation MediaPlayer QuartzCore Security
    ];
```

File: pkgs/applications/audio/psst/default.nix [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  src = fetchFromGitHub {
+    owner = "jpochyla";
+    repo = pname;
+    rev = "724aa251407dfa93b8ede8e879fd5ca35de1ddd2";
+    sha256 = "08m6h2gkgnpmfpw93zb6rhrvbw5rhq0mgpcfzingdvsqw531r13w";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [ pkg-config ];
+  buildInputs = [ gtk3 glib gdk-pixbuf cairo pango dbus ]
```
Comment (figsoda) [inline pkgs/applications/audio/psst/default.nix:17]: i think some or all of these are only needed for linux, but I'm not sure

**Review by figsoda — CHANGES_REQUESTED** [review]

File: pkgs/applications/audio/psst/default.nix [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+rustPlatform.buildRustPackage rec {
+  pname = "psst";
+  version = "unstable-2021-08-19";
+
+  src = fetchFromGitHub {
+    owner = "jpochyla";
+    repo = pname;
+    rev = "724aa251407dfa93b8ede8e879fd5ca35de1ddd2";
+    sha256 = "08m6h2gkgnpmfpw93zb6rhrvbw5rhq0mgpcfzingdvsqw531r13w";
+    fetchSubmodules = true;
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/psst/default.nix:13]: The package has no submodules.

File: pkgs/applications/audio/psst/default.nix [hunk 10]
```diff
@@ -0,0 +1,30 @@
+{ lib, rustPlatform, fetchFromGitHub, pkg-config, gtk3, glib,
+gdk-pixbuf, cairo , pango, dbus, stdenv, darwin ? null }:
```
Comment (SuperSandro2000) [inline pkgs/applications/audio/psst/default.nix:2]:
```suggestion
gdk-pixbuf, cairo , pango, dbus, stdenv }:
```

**Comment by matthias-t** [comment]
I'm no longer interested in this, feel free to take over.

**Comment by figsoda** [comment]
psst is now in nixpkgs
https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/applications/audio/psst/default.nix

**Pull request closed without merging**